File: src/main/javascript/tree/BinaryTreeLevelOrderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
const levelOrder = function (root) {
    if (root == null)
        return [];

    const levels = [];
    traverse(root, 0, levels)

    return levels;
};

const traverse = (node, level, levels) => {

    if (node == undefined)
        return;

    if (levels.length === level) {
        levels.push([]);
    }

    levels[level].push(node.val);

    if (node.left !== undefined) {
        traverse(node.left, level+1, levels)
    }
    if (node.right !== undefined) {
        traverse(node.right, level+1, levels)
    }
}

File: src/main/go/tree/binary_tree_level_order_traversal.go
package tree

func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	var levels [][]int

	traverse(root, &levels, 0)
	return levels
}

func traverse(node *TreeNode, levels *[][]int, level int) {

	if node == nil {
		return
	}

	if len(*levels) == level {
		*levels = append(*levels, []int{})
	}
	(*levels)[level] = append((*levels)[level], node.Val)

	if node.Left != nil {
		traverse(node.Left, levels, level+1)
	}

	if node.Right != nil {
		traverse(node.Right, levels, level+1)
	}
}
